torch/csrc/jit/mobile/nnc/context.h

```diff
@@ -0,0 +1,207 @@
+  c10::impl::GenericList run(const c10::impl::GenericList& inputs) const;
+
+  // The name of the function as specified in the model code.
+  c10::QualifiedName name() const {
+    return name_;
+  }
+
+  void set_name(const c10::QualifiedName& name) {
+    name_ = name;
+  }
+
+  // The unique id of the generated NNC kernel corresponding to the function.
+  const std::string& nnc_kernel_id() const {
+    return nnc_kernel_id_;
+  }
+
+  void set_nnc_kernel_id(const std::string& name) {
+    nnc_kernel_id_ = name;
+  }
+
+  // The parameters (e.g. weights / bias tensors) to be passed to the generated
+  // NNC kernel.
+  const std::vector<at::Tensor>& parameters() const {
+    return parameters_;
+  }
+
+  void set_parameters(const std::vector<at::Tensor>& parameters) {
+    parameters_ = parameters;
+  }
+
+  const std::vector<InputSpec>& input_specs() const {
+    return input_specs_;
+  }
+
+  void set_input_specs(const std::vector<InputSpec>& input_specs) {
+    input_specs_ = input_specs;
+  }
+
+  const std::vector<OutputSpec>& output_specs() const {
+    return output_specs_;
+  }
+
+  void set_output_spec(const std::vector<OutputSpec>& output_specs) {
+    output_specs_ = output_specs;
+  }
+
+  const MemoryPlan& memory_plan() const {
+    return memory_plan_;
+  }
+
+  void set_memory_plan(const MemoryPlan& memory_plan) {
+    memory_plan_ = memory_plan;
+  }
+
+ private:
+  void init_execution_state() const;
+
+  c10::QualifiedName name_;
+  std::string nnc_kernel_id_;
+  std::vector<at::Tensor> parameters_;
+  std::vector<InputSpec> input_specs_;
+  std::vector<OutputSpec> output_specs_;
+  MemoryPlan memory_plan_;
+  mutable std::unique_ptr<ExecutionState> execution_state_;
+};
+
+// CompilationUnit consists of a set of compiled NNC functions. It has a 1-1
+// correspondence with a `Module`.
+// It's similar as torch::jit::mobile::CompilationUnit.
+class TORCH_API CompilationUnit {
+ public:
+  CompilationUnit() = default;
+  CompilationUnit(const CompilationUnit&) = delete;
+  CompilationUnit(CompilationUnit&&) = default;
+  CompilationUnit& operator=(const CompilationUnit&) = delete;
+  CompilationUnit& operator=(CompilationUnit&&) = default;
+
+  // Deserialize from an IValue that is generated by the 'serialize()' method.
+  explicit CompilationUnit(const c10::IValue& value);
+
+  // Serialize all registered functions into an IValue. The IValue will be save
+  // into the compiled TorchScript model file ahead-of-time on the host, and
+  // will be deserialized at runtime on the target device.
+  C10_NODISCARD c10::IValue serialize() const;
+
+  // Execute a registered function.
+  C10_NODISCARD c10::impl::GenericList run(
+      const c10::QualifiedName& function_name,
+      const c10::impl::GenericList& inputs) const;
+
+  // Register a function to the compilation unit.
+  void register_function(std::unique_ptr<Function> fn);
+
+ private:
+  C10_NODISCARD Function* find_function(const c10::QualifiedName& qn) const;
```

Comment:
> For my own education, why is C10_NODISCARD needed in all these places?

I added these to get rid of the internal fbcode lint warnings...

I just did a search in the codebase and found it's not being widely used. I read the comments in the original PR #20324 - looks like the intention was to workaround some compiler error rather than lint warning. So maybe we should make this a convention?

cc: @ezyang @swolchok @smessmer who might know better.

Reply:
`[[nodiscard]]` is C++17 and we are C++14. Therefore use of it in our codebase is nonportable and we must use a macro. We can consider removing the macro when we move to C++17. 

If the question is whether or not we need the `[[nodiscard]]` in the first place, it is because many compilers have a lint that require you to mark something as nodiscard if there are no side effects. To shut up the lint, you have to add nodiscard in a lot of places. Honestly this is kind of goofy (you gotta do it EVERYWHERE) but the code is net better, so you might as well do it.